### PR TITLE
[components]: utilities: ymodem: Add ymodem send protocol and ry_sy function application.

### PR DESCRIPTION
**Pull request by Steven-LiuSF**
## 拉取/合并请求描述：(PR description)

[
Achieve the function of sending and receiving files via UART using YMODEM。

Use the command "ry [uart0]" to receive the file. Use the command "sy file_path [uart0]" to send the file. The console is used by default. You can choose any uart if you want.
The test uses the console to send and receive files including text, PDF, mp3 without errors, compared with the original file. (The 0x1A at the end of the file is added by the Ymodem protocol to simulate the ctrl Z, which is normal.) Other uart also passed the test. The test uses a baud rate of 115200 and stable transmission rate at 4kB / s. It has been found that an error will occur when the baud rate reaches 1.5M. 

Optimization direction:
(1) Add 1024 Byte data packet support.
(2) Support higher serial baud rate transmission.
]

以下的内容不应该在提交PR时的message修改，修改下述message，PR会被直接关闭。请在提交PR后，浏览器查看PR并对以下检查项逐项check，没问题后逐条在页面上打钩。
The following content must not be changed in submitted PR message. Otherwise, the PR will be closed immediately. After submitted PR, please use web browser to visit PR, and check items one by one, and ticked them if no problem.

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [x] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other style
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 本拉取/合并请求代码是高质量的 Code in this PR is of high quality

### CODE REVIEW

File: components/utilities/ymodem/ry_sy.c [hunk 1]
```diff
@@ -0,0 +1,241 @@
+/**
+  * Copyright (c) 2019 Fuzhou Rockchip Electronics Co., Ltd
+  *
+  * SPDX-License-Identifier: Apache-2.0
+  ******************************************************************************
```
Comment (BernardXiong) [inline components/utilities/ymodem/ry_sy.c]: Please use the standard style of RT-Thread header files to write copy right information. The copyright is OK, but others, please keep the style. Thanks.

File: components/utilities/ymodem/ry_sy.c [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    struct rym_ctx *ctx,
+    rt_uint8_t *buf,
+    rt_size_t len)
+{
+    struct custom_ctx *cctx = (struct custom_ctx *)ctx;
+    struct stat file_buf;
+    char insert_0 = '\0';
+    rt_err_t err;
+
+    cctx->fd = open(cctx->fpath, O_RDONLY);
```
Comment (BernardXiong) [inline components/utilities/ymodem/ry_sy.c]: It's related with file system, please add RT_USING_DFS for dependent.

File: components/utilities/ymodem/ymodem.c [hunk 3]
```diff
@@ -7,12 +7,17 @@
...
  * Date           Author       Notes
  * 2013-04-14     Grissiom     initial implementation
+ * 2019-12-09     Steven Liu   add YMODEM send protocol
  */
 
 #include <rthw.h>
 #include "ymodem.h"
 
-static const rt_uint16_t ccitt_table[256] = {
+#define CRC_TABLE
```
Comment (BernardXiong) [inline components/utilities/ymodem/ymodem.c]: If you want to make CRC table as an option, you can use YMODEM_USING_CRC_TABLE and place it in KConfig.

File: components/utilities/ymodem/ymodem.h [hunk 4]
```diff
@@ -138,7 +145,30 @@ struct rym_ctx
  * second.
  */
 rt_err_t rym_recv_on_device(struct rym_ctx *ctx, rt_device_t dev, rt_uint16_t oflag,
-        rym_callback on_begin, rym_callback on_data, rym_callback on_end,
-        int handshake_timeout);
+                            rym_callback on_begin, rym_callback on_data, rym_callback on_end,
+                            int handshake_timeout);
+
+/** send a file on device dev with ymodem session ctx.
```
Comment (BernardXiong) [inline components/utilities/ymodem/ymodem.h]: Please notice the comments of `/**`, which is a doxygen comments to generate documentation. However, the other of this file, there is no doxygen comments. Please keep style.

**Comment by Steven-LiuSF** [comment]
Due to improper operation, more than 5 commits were submitted. Submit after refresh

**Pull request closed without merging**